openpgp/keys.go

```diff
@@ -500,13 +520,26 @@ EachPacket:
 					// We really should warn that there was a failure here. Not raise an error
 					// since this really shouldn't be a fail-stop error.
 				}
-			} else if pkt.SigType == packet.SigTypeKeyRevocation {
-				// These revocations won't revoke UIDs as handled above, so lookout!
-				revocations = append(revocations, pkt)
+			} else if current != nil && pkt.SigType == packet.SigTypeIdentityRevocation {
+				if err = e.PrimaryKey.VerifyUserIdSignature(current.Name, e.PrimaryKey, pkt); err == nil {
+					current.Revocation = pkt
+
+					// Remove current from Identities, because the revocation signature
+					// might come after the certification signature.
+					delete(e.Identities, current.Name)
```

Comment:
Good point about `SelfSignature`. But for some keys it would have to go back and find another `SelfSignature`. So maybe lets not invalidate any uids, keep everything in `Identities` list but have the `Revocation` field in Identity is some code wanted to filter out revoked ones?